### PR TITLE
Update AOI Analyze/Model Size Threshold

### PR DESCRIPTION
**Pull request by arottersman**
## Overview

The continental RWD feature has made shapes that sneak past our AOI size threshold, but are still too large to analyze or run TR55 on more commonplace. To catch these shapes we're adjusting the threshold in three ways:

1. Lower the threshold to `75000 km2`

Testing on staging found this to be closer to the current max we can handle for non-diagonal shapes 

2. Test threshold on shape's extent's area instead of shape's area

While trying to find a more appropriate size threshold we realized the size of an AOI is not the most important factor in determining whether or not the shape will timeout. A more appropriate test is on the AOI's extent because the slowest part of analyze is fetching the land and soil tiles from S3, and the number of tiles fetched is dependent on the extent of the shape ([see this comment on the original issue for examples](https://github.com/WikiWatershed/model-my-watershed/issues/1746#issuecomment-292627146))

3. Set `spray.request-timeout`

We also discovered while timing `mmw-geoprocessing` that shapes would timeout at 40 sec regardless of setting the SJS `timeout` higher. This is because the `spray` timeout params default to 40 sec, and must also be set. This PR sets the `spray``idle-timeout` and `request-timeout` to 42 sec to match the backend.

Connects #1746 

### Demo

screen shot 2017-04-07 at 2 58 27 pm

screen shot 2017-04-10 at 10 42 36 am

### Notes

Most shapes that timeout their land and soil analyze tabs succeed on their animal counts, point sources, and water quality measures. With this in mind, should the threshold be set on TR55 only and still allow analyze to complete? 

I have left things as is for now because allowing analyze for shapes that can't do Land/Soil has the disadvantage of showing the user more timeout errors on the first, most prominent tabs. They might not even know the other tabs have succeeded.

## Testing Instructions

 * Pull this branch,  bundle
* Draw a few square AOI's until you've found a box about `75000 km2` large
     * Confirm there is a "Max area exceeded" type message
* Note where two diagonally opposite corners of that AOI lie
     * Draw a skinny snake to connect the two points you noted above and confirm the max area message pops up again noting a similar max area extent as the first AOI you drew (ie the extent of the diagonal line of the square should match the square)
    
To assess the new threshold you can try out these shapes on staging:

Should timeout ~`80,000 km2`
https://staging.app.wikiwatershed.org/project/769/scenario/1462

Should succeed ~`67,000 km2`
https://staging.app.wikiwatershed.org/project/767/scenario/1457

### CODE REVIEW

File: deployment/ansible/roles/model-my-watershed.spark-jobserver/templates/spark-jobserver.conf.j2 [hunk 1]
```diff
@@ -38,4 +38,9 @@ spark {
 # spray #
 #########
 
-spray.can.server.parsing.max-content-length = 250m
+spray.can.server {
+        parsing.max-content-length = 250m
+        # Match the backend's timeout set here src/mmw/apps/modeling/tasks.py
+        idle-timeout = 42s
```
Comment (rajadain) [inline deployment/ansible/roles/model-my-watershed.spark-jobserver/templates/spark-jobserver.conf.j2]: Took a bit of spelunking to figure out why SJS wouldn't start. Finally discovered that `idle-timeout` must always be greater than the value of `request-timeout`:

```
[JobServer-akka.actor.default-dispatcher-4] ERROR akka.actor.OneForOneStrategy - requirement failed: idle-timeout must be > request-timeout (if the latter is not 'infinite')
akka.actor.ActorInitializationException: exception during creation
	at akka.actor.ActorInitializationException$.apply(Actor.scala:166)
	at akka.actor.ActorCell.create(ActorCell.scala:596)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:456)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:263)
	at akka.dispatch.Mailbox.run(Mailbox.scala:219)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:397)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.IllegalArgumentException: requirement failed: idle-timeout must be > request-timeout (if the latter is not 'infinite')
	at scala.Predef$.require(Predef.scala:233)
	at spray.can.server.ServerSettings$Timeouts.<init>(ServerSettings.scala:75)
	at spray.can.server.ServerSettings$.fromSubConfig(ServerSettings.scala:86)
	at spray.can.server.ServerSettings$.fromSubConfig(ServerSettings.scala:60)
	at spray.util.SettingsCompanion.apply(SettingsCompanion.scala:50)
	at spray.util.SettingsCompanion$$anonfun$apply$1.apply(SettingsCompanion.scala:36)
	at scala.collection.MapLike$class.getOrElse(MapLike.scala:128)
	at scala.collection.AbstractMap.getOrElse(Map.scala:58)
	at spray.util.SettingsCompanion.apply(SettingsCompanion.scala:35)
	at spray.can.server.HttpListener$$anonfun$1.apply(HttpListener.scala:34)
	at spray.can.server.HttpListener$$anonfun$1.apply(HttpListener.scala:34)
	at scala.Option.getOrElse(Option.scala:120)
	at spray.can.server.HttpListener.<init>(HttpListener.scala:34)
	at spray.can.HttpManager.newHttpListener(HttpManager.scala:84)
	at spray.can.HttpManager$$anonfun$receive$1$$anonfun$applyOrElse$1.apply(HttpManager.scala:76)
	at spray.can.HttpManager$$anonfun$receive$1$$anonfun$applyOrElse$1.apply(HttpManager.scala:76)
	at akka.actor.TypedCreatorFunctionConsumer.produce(Props.scala:343)
	at akka.actor.Props.newActor(Props.scala:252)
	at akka.actor.ActorCell.newActor(ActorCell.scala:552)
	at akka.actor.ActorCell.create(ActorCell.scala:578)
	... 9 more
```

Updating it to `45s` did the trick.

Also consider using 2 spaces for indentation, as the rest of the file does. Optionally right align the values as shown in the documentation http://spray.io/documentation/1.2.4/spray-can/configuration/

```yaml
spray.can.server {
  parsing.max-content-length = 250m

  # Match the backend's timeout set here src/mmw/apps/modeling/tasks.py
  # idle-timeout _must be_ greater than request-timeout
  idle-timeout    = 45s
  request-timeout = 42s
}
```

File: src/mmw/js/src/draw/utils.js [hunk 2]
```diff
@@ -84,15 +85,39 @@ function cancelDrawing(map) {
     map.fire('draw:drawstop');
 }
 
-// Return shape area in km2.
-function shapeArea(shape) {
```
Comment (rajadain) [inline src/mmw/js/src/draw/utils.js]: I can imagine `shapeArea` being useful enough to keep around, but I suppose we can always bring it back when we need it.

File: src/mmw/js/src/draw/views.js [hunk 3]
```diff
@@ -89,9 +89,9 @@ function validateShape(polygon) {
         d.reject(errorMsg);
     } else if (!utils.isValidForAnalysis(polygon)) {
         var maxArea = utils.MAX_AREA.toLocaleString(),
-            selectedArea = Math.floor(utils.shapeArea(polygon)).toLocaleString(),
-            message = 'Sorry, the area you have delineated is too large ' +
-                      'to analyze or model. ' + selectedArea + ' km² were ' +
+            selectedBoundingBoxArea = Math.floor(utils.shapeBoundingBoxArea(polygon)).toLocaleString(),
+            message = 'Sorry, the extent of the area you have delineated is too large ' +
```
Comment (rajadain) [inline src/mmw/js/src/draw/views.js]: Consider replacing "extent" with "bounding box"

File: deployment/ansible/roles/model-my-watershed.spark-jobserver/templates/spark-jobserver.conf.j2 (hunk 1 above)
Comment (arottersman) [inline deployment/ansible/roles/model-my-watershed.spark-jobserver/templates/spark-jobserver.conf.j2]: Thanks for figuring this out and sorry for the trouble!

**Comment by arottersman** [comment]
> After fixing issues with Spark JobServer, I tried with a ~65K km² AoI and still get a timeout every time

The screenshot you posted appears to be from MMW running on your machine — I think we still expect staging and production to be able to handle larger AOIs. my `localhost` can usually only handle shapes around 10,000 km2.

**Comment by rajadain** [comment]
Do you know if there's a way to beef up to staging levels on my local? I can give the worker VM more RAM and CPU if it needs. But since the bottleneck seems to be downloading from S3, do you think the AWS environment is much faster at that than we are?

**Comment by arottersman** [comment]
The worker stats are:

| Model |	vCPU | Mem (GiB) | Networking Performance |
| ---- | --- | --- | --- |
| r4.large | 2 | 15.25 | Up to 10 Gigabit |

I'd imagine it could be the AWS env is faster at downloading from s3...running SJS locally had the calculation completing pretty quick already.

**Comment by arottersman** [comment]
Two fixups to address spray issues/styling, adding back in `shapeArea`, and extent --> bounding box.

I think we should base the threshold on staging, but am open to other ideas.

File: src/mmw/js/src/draw/utils.js [hunk 4]
```diff
@@ -84,15 +85,44 @@ function cancelDrawing(map) {
...
+    var shapeLatLngPoints = getGeoJsonLatLngs(shape),
+        latLngBounds = L.latLngBounds(shapeLatLngPoints),
+        boundingBox = [
+            latLngBounds.getWest(),
+            latLngBounds.getSouth(),
+            latLngBounds.getEast(),
+            latLngBounds.getNorth()
+        ],
+        boundingBoxPolygon = turfBboxPolygon(boundingBox);
+        return shapeArea(boundingBoxPolygon);
```
Comment (rajadain) [inline src/mmw/js/src/draw/utils.js]: 🎨  outdent this line one level. Also consider having a blank space between variable declarations and `return` statement.

File: src/mmw/js/src/draw/utils.js [hunk 5]
```diff
@@ -84,15 +85,44 @@ function cancelDrawing(map) {
...
+        return L.GeoJSON.coordsToLatLngs(coordinates, 2);
+    }
+    return null;
+}
+
 // Return shape area in km2.
 function shapeArea(shape) {
     return coreUtils.changeOfAreaUnits(turfArea(shape),
-            'm<sup>2</sup>', 'km<sup>2</sup>');
+        'm<sup>2</sup>', 'km<sup>2</sup>');
```
Comment (rajadain) [inline src/mmw/js/src/draw/utils.js]: 🎨  indenting this line will clean the diff.

File: src/mmw/js/src/draw/utils.js [hunk 6]
```diff
@@ -4,10 +4,11 @@ var $ = require('jquery'),
     L = require('leaflet'),
     _ = require('lodash'),
     turfArea = require('turf-area'),
+    turfBboxPolygon = require('turf-bbox-polygon'),
     coreUtils = require('../core/utils');
 
 // Keep in sync with src/api/main.py in rapid-watershed-delineation.
-var MAX_AREA = 112700; // About the size of a large state (in km^2)
+var MAX_AREA = 75000; // About the size of a large state (in km^2)
```
Comment (rajadain) [inline src/mmw/js/src/draw/utils.js]: ✏️  The average US state is 183,637 km². Consider changing this to "About the size of West Virginia"

File: src/mmw/js/src/draw/utils.js (hunk 4 above)
Comment (arottersman) [inline src/mmw/js/src/draw/utils.js]: Somewhat surprised the linter 🚨  didn't catch this

**Comment by arottersman** [comment]
I've fixed the formatting and squashed down the fixups. Thanks for all your help with this!

**Pull request merged**